python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,263 @@
+        # status_queue.put( (tester, clock(), subprocess_object) )
+        self.status_queue = Queue()
+
+    ## Run the command asynchronously
+    def run(self, tester, thread_lock):
+        return
+
+    ## process test results
+    def returnToTestHarness(self, tester, thread_lock):
+        return
+
+    # Return post run command from derived classes
+    def postCommand(self):
+        return
+
+    # Allow derived schedulers to initiate findAndRunTests again
+    def goAgain(self):
+        return
+
+    ## Returns a list of tests to run
+    def getTests(self):
+        return
+
+    # Notify derived schedulers of status change
+    def notifySchedulers(self, tester):
+        return
+
+    # Return boolean on output ready to be read
+    def checkOutputReady(self, tester):
+        return True
+
+    # Call back to TestHarness to closeFiles in the event a derived scheduler wants to exit abnormally
+    def cleanup(self, exit_code=0):
+        self.harness.closeFiles(exit_code)
+
+    # private methods to add a tester to a queue and assign a thread to it
+    # TODO: refactor runnerGo and statusGo
+    def _runnerGo(self, tester_name):
+        self.runner_queue.put(tester_name)
+        return self.runner_pool.apply_async(self.jobRunner, (self.runner_queue,))
+
+    # private methods to add a tester to a queue and assign a thread to it
+    # TODO: refactor runnerGo and statusGo
+    def _statusGo(self, tester_name):
+        self.status_queue.put(tester_name)
+        return self.status_pool.apply_async(self.statusRunner, (self.status_queue,))
+
+    # Loop through the testers and add them to the runner pool for execution
+    def schedule(self, testers):
+
+        # build a dictionary of testers so all threads have access to them
+        for tester in testers:
+            # { tester : testers } so that we know _this_ tester is related to these testers (dependency tracking)
+            self.scheduled_groups[tester.getTestName()] = testers
+
+            # { tester_name : tester_object } name to object map
+            self.scheduled_name_to_object[tester.getTestName()] = tester
+
+        # Add a test to the Runner Queue, and ask a thread to do some work (the thread pool has built in methods
+        # which prevent oversubscribing)
+        for tester in testers:
+            results = self._runnerGo(tester.getTestName())
+            # DEBUG this thread by using the get method. (Note: this is a blocking call)
+            # print results.get()
+
+    # Launch jobs stored in our queues. Method is blocked until both queue(s) are empty
+    def waitFinish(self):
+        while len(self.active_jobs) != 0 or not self.runner_queue.empty() or not self.status_queue.empty():
+            with self.thread_lock:
+                self.handleActiveTests()
+
+            # sleep for just a tick or two
+            sleep(0.1)
+
+        self.runner_pool.close()
+        self.runner_pool.join()
+        self.status_pool.close()
```

Comment:
Probably should have a comment here. I think we discussed that this had to be in this order.